### PR TITLE
(ADT-Diagnostics) Add data to the context for diagnostics.

### PR DESCRIPTION
**Pull request by azabbasi**

### CODE REVIEW

File: sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/DigitalTwinsAsyncClient.java [hunk 1]
```diff
@@ -55,11 +55,18 @@
 @ServiceClient(builder = DigitalTwinsClientBuilder.class, isAsync = true)
 public final class DigitalTwinsAsyncClient {
     private static final ClientLogger logger = new ClientLogger(DigitalTwinsAsyncClient.class);
-    private ObjectMapper mapper;
+    private final ObjectMapper mapper;
     private final DigitalTwinsServiceVersion serviceVersion;
     private final AzureDigitalTwinsAPIImpl protocolLayer;
     private static final Boolean includeModelDefinitionOnGet = true;
     private final JsonSerializer serializer;
+    private static final String DIGITAL_TWINS_TRACING_NAMESPACE_VALUE = "Microsoft.DigitalTwins";
```
Comment (abhipsaMisra) [inline sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/DigitalTwinsAsyncClient.java:63]: Q- is this something that is assigned to us? any reason it is "Microsoft.DigitalTwins" and not "Azure.DigitalTwins' (similar to our package name having "Azure" in it, and not "Microsoft")?

File: sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/DigitalTwinsAsyncClient.java [hunk 2]
```diff
@@ -151,7 +158,12 @@ public DigitalTwinsServiceVersion getServiceVersion() {
     <T> Mono<DigitalTwinsResponse<T>> createOrReplaceDigitalTwinWithResponse(String digitalTwinId, T digitalTwin, Class<T> clazz, CreateOrReplaceDigitalTwinOptions options, Context context) {
         return protocolLayer
             .getDigitalTwins()
-            .addWithResponseAsync(digitalTwinId, digitalTwin, OptionsConverter.toProtocolLayerOptions(options), context)
+            .addWithResponseAsync(
+                digitalTwinId,
+                digitalTwin,
+                OptionsConverter.toProtocolLayerOptions(options),
+                context.addData(AZ_TRACING_NAMESPACE_KEY, DIGITAL_TWINS_TRACING_NAMESPACE_VALUE)
```
Comment (abhipsaMisra) [inline sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/DigitalTwinsAsyncClient.java:165]: cool!

File: sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/DigitalTwinsAsyncClient.java (hunk 1 above)
Comment (azabbasi) [reply sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/DigitalTwinsAsyncClient.java:63]: The .NET diagnostics required the resource provider namespace, so I stayed consistent here.

**Review by abhipsaMisra — APPROVED** [review]
Q - Is this the same as the data that we trace in .NET?

**Review by barustum — APPROVED** [review]

File: sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/DigitalTwinsAsyncClient.java (hunk 1 above)
Comment (srnagar) [reply sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/DigitalTwinsAsyncClient.java:63]: @abhipsaMisra The list of resource provider names can be found here - https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/azure-services-resource-providers

Looks like `Microsoft.DigitalTwins` is the right namespace to use.

File: sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/DigitalTwinsAsyncClient.java [hunk 3]
```diff
@@ -392,7 +417,14 @@ public DigitalTwinsServiceVersion getServiceVersion() {
...
             .getDigitalTwins()
-            .addRelationshipWithResponseAsync(digitalTwinId, relationshipId, relationship, OptionsConverter.toProtocolLayerOptions(options), context)
+            .addRelationshipWithResponseAsync(
+                digitalTwinId,
+                relationshipId,
+                relationship,
+                OptionsConverter.toProtocolLayerOptions(options),
+                context.addData(AZ_TRACING_NAMESPACE_KEY, DIGITAL_TWINS_TRACING_NAMESPACE_VALUE)
+                    .addData(DIGITAL_TWIN_ID_TRACING_KEY, digitalTwinId)
+                    .addData(RELATIONSHIP_ID_TRACING_KEY, relationshipId))
```
Comment (srnagar) [inline sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/DigitalTwinsAsyncClient.java:427]: Where are these keys read? `AZ_TRACING_NAMESPACE_KEY` is a known azure-core key that is read by the `OpenTelemetryHttpPolicy`.  

Did you mean to add `digitalTwinId` and `relationshipId` as tracing span attributes?

cc: @samvaity

**Comment by azabbasi** [comment]
> Q - Is this the same as the data that we trace in .NET?

YES

File: sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/DigitalTwinsAsyncClient.java (hunk 2 above)
Comment (bikamani) [inline sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/DigitalTwinsAsyncClient.java:165]: what does this do?
what is AZ_TRACING_NAMESPACE_KEY?

File: sdk/digitaltwins/azure-digitaltwins-core/src/test/java/com/azure/digitaltwins/core/QueryTests.java [hunk 4]
```diff
@@ -29,7 +29,7 @@
     @Override
     public void validQuerySucceeds(HttpClient httpClient, DigitalTwinsServiceVersion serviceVersion) {
         DigitalTwinsClient client = getClient(httpClient, serviceVersion);
-        int pageSize = 5;
+        int pageSize = 3;
```
Comment (bikamani) [inline sdk/digitaltwins/azure-digitaltwins-core/src/test/java/com/azure/digitaltwins/core/QueryTests.java:32]: why is this changed?

File: sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/DigitalTwinsAsyncClient.java (hunk 2 above)
Comment (azabbasi) [reply sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/DigitalTwinsAsyncClient.java:165]: it's the agreed upon key for the key-value pair for namespace in the context.

File: sdk/digitaltwins/azure-digitaltwins-core/src/test/java/com/azure/digitaltwins/core/QueryTests.java (hunk 4 above)
Comment (azabbasi) [reply sdk/digitaltwins/azure-digitaltwins-core/src/test/java/com/azure/digitaltwins/core/QueryTests.java:32]: this change doesn't really matter, I changed it during fixing this test.

Comment (azabbasi) [reply sdk/digitaltwins/azure-digitaltwins-core/src/test/java/com/azure/digitaltwins/core/QueryTests.java:32]: it just changes the max-item-per-page to 3 from 5 ...

File: sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/DigitalTwinsAsyncClient.java (hunk 2 above)
Comment (azabbasi) [reply sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/DigitalTwinsAsyncClient.java:165]: This adds context to the diagnostic tracing for each API call.

File: sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/DigitalTwinsAsyncClient.java (hunk 3 above)
Comment (samvaity) [reply sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/DigitalTwinsAsyncClient.java:427]: Yes, as @srnagar said,  `DIGITAL_TWIN_ID_TRACING_KEY` and `RELATIONSHIP_ID_TRACING_KEY` are not identified keys on the tracing end(azure-core-tracing-opetelemetry). 

And hence, these attributes would not be attached to the tracing span until it's recognized and supported in the `azure-core-tracing-opetelemetry` library itself

Comment (azabbasi) [reply sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/DigitalTwinsAsyncClient.java:427]: I see, I will remove all additional keys since they are not supported.
